File: study/lua/src/app/scene/view/seasonSport/PopupSuspendTimeView.lua
-- @Author panhoa
-- @Date 10.9.2018
-- @Role

local PopupBase = require("app.ui.PopupBase")
local PopupSuspendTimeView = class("PopupSuspendTimeView", PopupBase)
local scheduler = require("cocos.framework.scheduler")


function PopupSuspendTimeView:ctor()
    self._commonNodeBk  = nil
    self._btnBack       = nil
    self._textHint      = nil

	local resource = {
		file = Path.getCSB("PopupSuspendTimeView", "seasonSport"),
		binding = {
			_btnBack = {
				events = {{event = "touch", method = "_onBtnBack"}}
			},
		}
	}
    PopupSuspendTimeView.super.ctor(self, resource, false, false)
end

-- @Role    自定信息
-- @Param   strTitle 标题
-- @Param   strContent 内容
-- @Param   strButton  按钮名
function PopupSuspendTimeView:setCustomText(strTitle, strContent, strButton, strContentEnd, offsetY)
    self._commonNodeBk:setTitle(strTitle)
    
    self._nodeDesc:removeAllChildren()
    local richText = ccui.RichText:createRichTextByFormatString(strContent,
    {defaultColor = Colors.BRIGHT_BG_TWO, defaultSize = 22, other ={
        [1] = {fontSize = 22}
    }})
    self._nodeDesc:addChild(richText)

    if strContentEnd ~= nil and strContentEnd ~= "" then
        local richText2 = ccui.RichText:createRichTextByFormatString(strContentEnd,
        {defaultColor = Colors.BRIGHT_BG_TWO, defaultSize = 22, other ={
            [1] = {fontSize = 22}
        }})
        richText2:setPositionY(richText:getPositionY() - 30)
        self._nodeDesc:addChild(richText2)    
    end
    self._nodeDesc:setPositionY(self._oriPositionY - offsetY)
    self._btnBack:setString(strButton)
end

function PopupSuspendTimeView:onCreate()
    self._commonNodeBk:addCloseEventListener(handler(self, self._onCloseBack))
    self._oriPositionY = self._nodeDesc:getPositionY()
end

function PopupSuspendTimeView:onEnter()
    local suspendTime = G_UserData:getSeasonSport():getSuspendTime()
    if tonumber(G_ServerTime:getLeftSeconds(suspendTime)) > 0 then
        self._textHint:setString(G_ServerTime:getLeftSecondsString(suspendTime, "00：00：00"))
        self._countDownScheduler = scheduler.scheduleGlobal(handler(self, self._update), 0.5)
    end
    self._image_time:setVisible(tonumber(G_ServerTime:getLeftSeconds(suspendTime)) > 0)
end

function PopupSuspendTimeView:onExit()
    if self._countDownScheduler then
        scheduler.unscheduleGlobal(self._countDownScheduler)
    end
	self._countDownScheduler = nil
end

function PopupSuspendTimeView:setCloseCallBack(closeCallBack)
    self._closeCallBack = closeCallBack
end

function PopupSuspendTimeView:setOkCallBack(okCallBack)
    self._okCallBack = okCallBack
end

function PopupSuspendTimeView:_onCloseBack(sender)
    if self._closeCallBack then
        self._closeCallBack()
    end
	self:close()
end

function PopupSuspendTimeView:_onBtnBack(sender)
    if self._closeCallBack then
        self._closeCallBack()
    end

    if self._okCallBack then
        self._okCallBack()
    end
	self:close()
end

-- @Role 	Update
function PopupSuspendTimeView:_update(dt)
    local suspendTime = G_UserData:getSeasonSport():getSuspendTime()
    if tonumber(G_ServerTime:getLeftSeconds(suspendTime)) > 0 then
        self._textHint:setString(G_ServerTime:getLeftSecondsString(suspendTime, "00：00：00"))
    else
        if self._closeCallBack then
            self._closeCallBack()
        end
        self:close()
    end
end


return PopupSuspendTimeView

File: study/ts/script/scene/view/seasonSport/PopupSuspendTimeView.ts
const { ccclass, property } = cc._decorator;

import CommonButtonLevel0Highlight from '../../../ui/component/CommonButtonLevel0Highlight'

import CommonNormalMiniPop from '../../../ui/component/CommonNormalMiniPop'
import { RichTextExtend } from '../../../extends/RichTextExtend';
import { Colors, G_UserData, G_ServerTime } from '../../../init';
import { handler } from '../../../utils/handler';
import PopupBase from '../../../ui/PopupBase';

@ccclass
export default class PopupSuspendTimeView extends PopupBase {

    @property({ type: CommonNormalMiniPop, visible: true })
    _commonNodeBk: CommonNormalMiniPop = null;

    @property({ type: cc.Node, visible: true })
    _nodeDesc: cc.Node = null;

    @property({ type: cc.Sprite, visible: true })
    _image_time: cc.Sprite = null;

    @property({ type: cc.Label, visible: true })
    _textHint: cc.Label = null;

    @property({ type: CommonButtonLevel0Highlight, visible: true })
    _btnBack: CommonButtonLevel0Highlight = null;

    private _oriPositionY;
    private _closeCallBack;
    private _okCallBack;

    public setCustomText(strTitle, strContent, strButton, strContentEnd, offsetY) {
        this._commonNodeBk.setTitle(strTitle);
        this._nodeDesc.removeAllChildren();
        var richText = RichTextExtend.createRichTextByFormatString(strContent, {
            defaultColor: Colors.BRIGHT_BG_TWO,
            defaultSize: 22,
            other: { 1: { fontSize: 22 } }
        });
        this._nodeDesc.addChild(richText.node);
        if (strContentEnd != null && strContentEnd != '') {
            var richText2 = RichTextExtend.createRichTextByFormatString(strContentEnd, {
                defaultColor: Colors.BRIGHT_BG_TWO,
                defaultSize: 22,
                other: { 1: { fontSize: 22 } }
            });
            richText2.node.y = richText.node.y - 30;
            this._nodeDesc.addChild(richText2.node);
        }
        if (this._oriPositionY == null) {
            this._oriPositionY = this._nodeDesc.y;
        }
        this._nodeDesc.y = (this._oriPositionY - offsetY);
        this._btnBack.setString(strButton);
    }

    public onCreate() {
        this._commonNodeBk.addCloseEventListener(handler(this, this._onCloseBack));
        this._btnBack.addClickEventListenerEx(handler(this, this._onBtnBack));
    }

    public onEnter() {
        var suspendTime = G_UserData.getSeasonSport().getSuspendTime();
        if ((G_ServerTime.getLeftSeconds(suspendTime)) > 0) {
            this._textHint.string = G_ServerTime.getLeftSecondsString(suspendTime, '00：00：00');
            this.schedule(handler(this, this._update), 0.5);
        }
        this._image_time.node.active = (G_ServerTime.getLeftSeconds(suspendTime)) > 0;
    }

    public onExit() {
        this.unschedule(handler(this, this._update));
    }

    public setCloseCallBack(closeCallBack) {
        this._closeCallBack = closeCallBack;
    }

    public setOkCallBack(okCallBack) {
        this._okCallBack = okCallBack;
    }

    private _onCloseBack(sender) {
        if (this._closeCallBack) {
            this._closeCallBack();
        }
        this.close();
    }

    private _onBtnBack(sender) {
        if (this._closeCallBack) {
            this._closeCallBack();
        }
        if (this._okCallBack) {
            this._okCallBack();
        }
        this.close();
    }

    private _update(dt) {
        var suspendTime = G_UserData.getSeasonSport().getSuspendTime();
        if (G_ServerTime.getLeftSeconds(suspendTime) > 0) {
            this._textHint.string = G_ServerTime.getLeftSecondsString(suspendTime, '00：00：00');
        } else {
            if (this._closeCallBack) {
                this._closeCallBack();
            }
            this.close();
        }
    }
}